embulk-jruby-extension/src/main/java/org/jruby/util/RubyDateParser.java

```diff
@@ -0,0 +1,605 @@
+    static {
+        numPatterns = new HashSet<>();
+        numPatterns.add(Format.FORMAT_CENTURY); // 'C'
+        // D
+        numPatterns.add(Format.FORMAT_DAY); // 'd'
+        numPatterns.add(Format.FORMAT_DAY_S); // 'e'
+        // F
+        numPatterns.add(Format.FORMAT_WEEKYEAR); // 'G'
+        numPatterns.add(Format.FORMAT_WEEKYEAR_SHORT); // 'g'
+        numPatterns.add(Format.FORMAT_HOUR); // 'H'
+        numPatterns.add(Format.FORMAT_HOUR_M); // 'I'
+        numPatterns.add(Format.FORMAT_DAY_YEAR); // 'j'
+        numPatterns.add(Format.FORMAT_HOUR_BLANK); // 'k'
+        numPatterns.add(Format.FORMAT_MILLISEC); // 'L'
+        numPatterns.add(Format.FORMAT_HOUR_S); // 'l'
+        numPatterns.add(Format.FORMAT_MINUTES); // 'M'
+        numPatterns.add(Format.FORMAT_MONTH); // 'm'
+        numPatterns.add(Format.FORMAT_NANOSEC); // 'N'
+        // Q, R, r
+        numPatterns.add(Format.FORMAT_SECONDS); // 'S'
+        numPatterns.add(Format.FORMAT_EPOCH); // 's'
+        // T
+        numPatterns.add(Format.FORMAT_WEEK_YEAR_S); // 'U'
+        numPatterns.add(Format.FORMAT_DAY_WEEK2); // 'u'
+        numPatterns.add(Format.FORMAT_WEEK_WEEKYEAR); // 'V'
+        // v
+        numPatterns.add(Format.FORMAT_WEEK_YEAR_M); // 'W'
+        numPatterns.add(Format.FORMAT_DAY_WEEK); // 'w'
+        // X, x
+        numPatterns.add(Format.FORMAT_YEAR_LONG); // 'Y'
+        numPatterns.add(Format.FORMAT_YEAR_SHORT); // 'y'
+    }
+
+    private static boolean matchAtNumPatterns(Token token) // NUM_PATTERN_P
+    {
+        if (token == null) {
+            return false;
+        }
+
+        Format f = token.getFormat();
+        if (f == Format.FORMAT_STRING && isDigit(((String)token.getData()).charAt(0))) {
+            return true;
+
+        } else if (numPatterns.contains(f)) {
+            return true;
+
+        }
+        return false;
+    }
+
+    private final ThreadContext context;
+    // Use RubyDateFormatter temporarily because it has useful lexer, token and format types
+    private final RubyDateFormatter dateFormat;
+
+    private List<Token> compiledPattern;
+    private int pos;
+    private String text;
+    private ParsedValues values;
+
+    public RubyDateParser()
+    {
+        this(Ruby.newInstance().getCurrentContext());
+    }
+
+    public RubyDateParser(ThreadContext context)
+    {
+        this.context = context;
+        this.dateFormat = new RubyDateFormatter(context);
+    }
+
+    public List<Token> compilePattern(String format)
+    {
+        return compilePattern(context.runtime.newString(format), true);
+    }
+
+    public List<Token> compilePattern(RubyString format, boolean dateLibrary)
+    {
+        return dateFormat.compilePattern(format, dateLibrary);
+    }
+
+    private static Token getToken(List<Token> compiledPattern, int index)
+    {
+        return compiledPattern.get(index);
+    }
+
+    private static Token nextToken(List<Token> compiledPattern, int index)
```

Comment:
This method is used only with matchAtNumPatterns. Why don't you combine them to one method?